coreblocks/func_blocks/fu/unsigned_multiplication/iterative_multiplication.py

```diff
@@ -0,0 +1,146 @@
+        self.number_of_multiplications = self.number_of_chunks_multiplication**2
+        self.number_of_steps = math.ceil(self.number_of_multiplications / self.dsp_number)
+        self.result_lvl = math.ceil(math.log2(self.number_of_chunks))
+
+        self.ready = Signal()
+        self.issue = Signal()
+        self.step = Signal(range(self.number_of_steps + 1), reset=self.number_of_steps)
+        self.i1 = Signal(self.n_padding)
+        self.i2 = Signal(self.n_padding)
+        self.result = Signal(2 * n)
+        self.valid = Signal()
+        self.getting_result = Signal()
+
+        self.values_array = [
+            [
+                [Signal(self.dsp_width * (1 << i) * 2) for _ in range(self.number_of_chunks >> i)]
+                for _ in range(self.number_of_chunks >> i)
+            ]
+            for i in range(self.result_lvl + 1)
+        ]
+        self.valid_array = [Signal() for _ in range(self.result_lvl + 1)]
+
+    def elaborate(self, platform=None):
+        m = TModule()
+
+        self.dsp_units = []
+        for i in range(self.dsp_number):
+            unit = DSPMulUnit(self.dsp_width)
+            self.dsp_units.append(unit)
+            setattr(m.submodules, f"dsp_unit_{i}", unit)
+
+        with m.If((self.step < (self.number_of_steps - 1)) | (self.valid & ~self.getting_result)):
+            m.d.comb += self.ready.eq(0)
+        with m.Else():
+            m.d.comb += self.ready.eq(1)
+
+        m.d.comb += self.result.eq(self.values_array[self.result_lvl][0][0])
+        m.d.comb += self.valid.eq(self.valid_array[self.result_lvl])
+
+        with m.If(~self.valid | self.getting_result):
+            for i in range(1, self.result_lvl + 1):
+                m.d.sync += self.valid_array[i].eq(self.valid_array[i - 1])
+
+            with m.If(self.step == self.number_of_steps - 1):
+                m.d.sync += self.valid_array[0].eq(1)
+            with m.Else():
+                m.d.sync += self.valid_array[0].eq(0)
+
+            with m.If(self.step < self.number_of_steps):
+                m.d.sync += self.step.eq(self.step + 1)
+
+            with m.If(self.issue):
+                m.d.sync += self.step.eq(0)
+
+            dsp_idx = 0
+            for i in range(self.number_of_multiplications):
+                a = i // self.number_of_chunks_multiplication
+                b = i % self.number_of_chunks_multiplication
+                chunk_i1 = self.i1[a * self.dsp_width : (a + 1) * self.dsp_width]
+                chunk_i2 = self.i2[b * self.dsp_width : (b + 1) * self.dsp_width]
+                with m.If((i >= self.step * self.dsp_number) & (i < (self.step + 1) * self.dsp_number)):
+                    dsp_idx = (dsp_idx + 1) % self.dsp_number
+                    with Transaction().body(m):
+                        res = self.dsp_units[dsp_idx].compute(m, i1=chunk_i1, i2=chunk_i2)
+                        m.d.sync += self.values_array[0][a][b].eq(res)
```

Comment:
```suggestion
                run_on_step = i // self.dsp_number
                dsp_idx = i % self.dsp_number

                with Transaction().body(m, request=(self.step == run_on_step)):
                    res = self.dsp_units[dsp_idx].compute(m, i1=chunk_i1, i2=chunk_i2)
                    m.d.sync += self.values_array[0][a][b].eq(res)

```

I would suggest to simplify and clean this logic as provided:
Python `dsp_idx` assignment is not affected by Amaranth `m.If()`, so it is effectively only disabling Transaction (see `request` signal).
Values can be optimized to constants resolved  at generation, also it easier to understand (at least for me).

Please verify if I correctly understood the logic and that suggested change is equivalent.

Reply:
Yes, your code is equivalent and better looking. I will make this change.